server/src/main/java/org/elasticsearch/index/mapper/PostParsePhase.java

```diff
@@ -0,0 +1,484 @@
+        }
+
+        @Override
+        public SortedSetDocValues getSortedSetDocValues(String field) throws IOException {
+            List<BytesRef> values = document.getFields().stream()
+                .filter(f -> Objects.equals(f.name(), field))
+                .filter(f -> f.fieldType().docValuesType() == DocValuesType.BINARY)
+                .map(IndexableField::binaryValue)
+                .sorted()
+                .collect(Collectors.toList());
+            return sortedSetDocValues(values);
+        }
+
+        @Override
+        public FieldInfos getFieldInfos() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void document(int docID, StoredFieldVisitor visitor) throws IOException {
+            visitor.binaryField(SOURCE_FIELD_INFO, sourceBytes.toBytesRef().bytes);
+        }
+
+        @Override
+        public CacheHelper getCoreCacheHelper() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public Terms terms(String field) throws IOException {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public NumericDocValues getNormValues(String field) throws IOException {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public Bits getLiveDocs() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public PointValues getPointValues(String field) throws IOException {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void checkIntegrity() throws IOException {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public LeafMetaData getMetaData() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public Fields getTermVectors(int docID) throws IOException {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public int numDocs() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public int maxDoc() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        protected void doClose() throws IOException {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public CacheHelper getReaderCacheHelper() {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    private static final FieldInfo SOURCE_FIELD_INFO = new FieldInfo(
+        "_source",
+        0,
+        false,
+        false,
+        false,
+        IndexOptions.NONE,
+        DocValuesType.NONE,
+        -1,
+        Collections.emptyMap(),
+        0,
+        0,
+        0,
+        false
+    );
+
+    private static NumericDocValues numericDocValues(List<Number> values) {
```

Comment:
I didn't really like the `MemoryIndex`, but one of the advantages was that it'd enforce the same sort of rounding as you'd get if you loaded from doc values. I *think* this keeps that rounding, but I imagine it'll be a thing we'd have to have a clutch of tests for. If we don't then sneaky things'll happen, I think.

Reply:
Everything here is a long - `NumericDocValuesField` takes longs, and `NumericDocValues` returns them, so all the rounding and double encoding happens at the mapper level.  I'm going to add some more comprehensive tests directly on the lazy reader which should reassure everybody :)

Reply:
Say your `long` references a `half_float` field. One we'd build doc values for. If your long were a runtime field it'd get the `half_float` precision. But if its a calculated field will it get `double` precision?

Reply:
The `half_float` field will have added its values to the underlying `Document` by calling `fields.add(new SortedNumericDocValuesField(name, HalfFloatPoint.halfFloatToSortableShort(value.floatValue())));` (see https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java#L206) which encodes things into a short value, but then casts them to a long for docvalues.  Then when we retrieve the doc values in fielddata builder, things are passed through `SortedNumericIndexFieldData#SingleHalfFloatValues` which reads longs from the underlying NumericDocValues and decodes them back to floats via `HalfFloatPoint.sortableShortToHalfFloat((short) in.longValue());`. Precision doesn't matter for doc values, everything is always stored as `long`.

Reply:
Those are the sorts of tests that I was worried about. Making sure we lose precision in the same way. It looks like we do *now* but one day someone is going to break it if we don't have a test for it.

Reply:
Yeah I want to add some tests to MapperTestCase that check for every field type that you can retrieve these values at index time and a search time and they behave the same way, but I think we should leave that to a follow up because it will blow up this PR a lot if we try and add it here.

Reply:
Also: the MemoryIndex just stores things as longs as well, so doing this via MI or via these hand-rolled DV implementations doesn't make any difference to what we need to test.

Reply:
:+1: on MemoryIndex.

I don't mind blowing up the PR if it proves that it can handle sneaky stuff. Maybe just a single test case for something sneaky like `half_float` or `float`.